src/lib/work_partitioner/work_partitioner.ml

```diff
@@ -1 +1,240 @@
+      Zkapp_command_job_pool.job ) =
+  let%map.Option spec = Pending_zkapp_command.next_job_spec pending in
+  let job_id =
+    Work.Id.Sub_zkapp.of_single
+      ~job_id:(Id_generator.next_id partitioner.id_generator ())
+      zkapp_id
+  in
+  let job =
+    Work.With_job_meta.
+      { spec; job_id; issued_since_unix_epoch = epoch_now (); sok_message }
+  in
+  assert (
+    phys_equal `Ok
+      (Sent_zkapp_job_pool.add ~id:job_id ~job
+         partitioner.zkapp_jobs_sent_by_partitioner ) ) ;
+
+  Work.Spec.Partitioned.Poly.Sub_zkapp_command { job; data = () }
+
+let issue_from_zkapp_command_work_pool ~(partitioner : t) () :
+    Work.Spec.Partitioned.Stable.Latest.t option =
+  Zkapp_command_job_pool.iter_until
+    ~f:(issue_from_pending_zkapp_command ~partitioner)
+    partitioner.zkapp_command_jobs
+
+let convert_single_work_from_selector ~(partitioner : t)
+    ~(single_spec : Work.Spec.Single.t) ~sok_message ~pairing :
+    Work.Spec.Partitioned.Stable.Latest.t =
+  let job =
+    Work.With_job_meta.
+      { spec = single_spec
+      ; job_id = pairing
+      ; issued_since_unix_epoch = epoch_now ()
+      ; sok_message
+      }
+  in
+  match single_spec with
+  | Transition (input, witness) -> (
+      match witness.transaction with
+      | Command (Zkapp_command zkapp_command) -> (
+          let witness = Transaction_witness.read_all_proofs_from_disk witness in
+          match
+            Snark_worker_shared.extract_zkapp_segment_works
+              ~m:partitioner.transaction_snark ~input ~witness ~zkapp_command
+          with
+          | Ok (_ :: _ as all) ->
+              let unscheduled_segments =
+                all
+                |> List.map ~f:(fun (witness, spec, statement) ->
+                       Work.Spec.Sub_zkapp.Segment { statement; witness; spec }
+                       |> Work.Spec.Sub_zkapp.read_all_proofs_from_disk )
+                |> Queue.of_list
+              in
+              let pending_zkapp_command =
+                Pending_zkapp_command.create ~job ~unscheduled_segments
+              in
+              let pending_zkapp_command_job =
+                Work.With_job_meta.
+                  { spec = pending_zkapp_command
+                  ; job_id = pairing
+                  ; issued_since_unix_epoch = epoch_now ()
+                  ; sok_message
+                  }
+              in
+              assert (
+                phys_equal `Ok
+                  (Zkapp_command_job_pool.add ~id:pairing
+                     ~job:pending_zkapp_command_job
+                     partitioner.zkapp_command_jobs ) ) ;
+              issue_from_pending_zkapp_command ~partitioner
+                pending_zkapp_command_job
+              |> Option.value_exn
+                   ~message:
+                     "FATAL: we already inserted work into partitioner so this \
+                      shouldn't happen"
+          | Ok [] ->
+              failwith "No witness generated"
```

Comment:
Should we propagate this error?

Reply:
This error won't happen, but I don't like it's leaked in this context. Let's have a `Snark_worker_shared.extract_zkapp_segment_works_single` function that will take exactly one zkapp tx and non-empty list.